src/MatrixTypes.f90

```diff
@@ -912,28 +872,47 @@ SUBROUTINE matvec_DistrBandedMatrixType(thisMatrix,x,y,t,ul,d,incx,a,b)
 !> @param req The array of active requests for data
 !> @param idxReq The array of requests for indices
 !>
-SUBROUTINE pop_recv(acc,valBuf,idxBuf,ctBuf,idx,req,idxReq)
+SUBROUTINE pop_recv(acc,valBuf,thisMat,ctBuf,idx,req,f)
   REAL(SRK), INTENT(INOUT) :: acc(:)
   REAL(SRK), INTENT(INOUT) :: valBuf(:,:)
+  CLASS(DistributedBandedMatrixType),INTENT(INOUT) :: thisMat
   INTEGER(SIK), INTENT(INOUT) :: ctBuf(MATVEC_SLOTS)
-  INTEGER(SIK), INTENT(INOUT) :: idxBuf(:,:)
-  INTEGER(SIK), INTENT(IN) :: idx
+  INTEGER(SIK), INTENT(INOUT) :: idx
   INTEGER(SIK), INTENT(INOUT) :: req(MATVEC_SLOTS)
-  INTEGER(SIK), INTENT(INOUT) :: idxReq(MATVEC_SLOTS)
-  INTEGER(SIK) :: mpierr
+  LOGICAL(SBK),OPTIONAL,INTENT(IN) :: f
+  INTEGER(SIK) :: stt,stp,mpierr,i,rank
+  LOGICAL(SBK) :: force
+  CALL MPI_Comm_Rank(thisMat%comm,rank,mpierr)
 
-  IF(req(idx) == 0 .AND. idxReq(idx) == 0) RETURN
+  force=.FALSE.
+  IF (PRESENT(f)) force=f
+  IF (force) THEN
+    IF (req(idx) == MPI_REQUEST_NULL) RETURN
+    CALL MPI_Wait(req(idx),MPI_STATUS_IGNORE,mpierr)
+  ELSE
+    IF (ANY(req == MPI_REQUEST_NULL)) THEN
+      DO i=1,SIZE(req)
+        IF (req(i) == MPI_REQUEST_NULL) THEN
+          idx=i
+          RETURN
```

Comment:
I'm having a hard time understanding why this part of the IF-block `ANY(req == MPI_REQUEST_NULL)=TRUE` is needed.

Reply:
The ANY(req == ...) is what controls whether the code searches for the empty slot (to return to the matrix multiplication code) or waits on ongoing communication to finish.
If any are MPI_REQUEST_NULL then that is an empty buffer and communication will start to that buffer.

I know we discussed this in our call but I'm putting it here if in the future, anyone has the same question.